File: i-payout/recipes/_old/python/swift-wire.py
import base64
import requests
import os
from dotenv import load_dotenv

load_dotenv()

base_url = "https://merchantapi.testewallet.com/api/v1"

username = os.environ.get("IPAYOUT_USERNAME")
password = os.environ.get("IPAYOUT_PASSWORD")
merchant_id = os.environ.get("IPAYOUT_MERCHANT_ID")

auth_str = f"{username}:{password}"
encoded_auth_str = base64.b64encode(auth_str.encode()).decode()

token_url = f"{base_url}/authentication/login"
headers = {
    "accept": "application/json",
    "authorization": f"Basic {encoded_auth_str}",
    "X-MerchantId": merchant_id,
}

response = requests.get(token_url, headers=headers)
api_token = response.json()["data"]["token"]


customer_url = f"{base_url}/beneficiaries/create"
body = {
    "username": "john_doe",
    "firstName": "John",
    "lastName": "Doe",
    "emailAddress": "johndoe@mail.com",
}
headers = {
    "Authorization": f"Bearer {api_token}",
    "X-MerchantId": merchant_id,
    "accept": "application/json",
    "content-type": "application/*+json",
}

response = requests.post(customer_url, headers=headers, json=body)

# customer_token = response.json()["data"]["beneficiaryToken"]
customer_token = "e6252377-e999-46fd-bb70-3c4bdb83edf8"

transfer_method_url = f"{base_url}/transfermethod/beneficiaries/{customer_token}/wire"
body = {
    "accountNickName": "John US Wire Account",
    "beneficiaryCountry": "US",
    "accountType1": "personal",
    "routingNumber": "12345",
    "beneficiaryAddress1": "123 Elm St",
    "beneficiaryCity": "Los Angeles",
    "beneficiaryState": "CA",
    "beneficiaryZipCode": "90001",
    "bankCountry": "US",
    "accountNumber": "987654321",
    "bankName": "Bank of America",
    "bankAddress1": "123 Bank St",
    "bankCity": "New York",
    "bankState": "New York",
    "bankZipCode": "10001",
}

headers = {
    "Authorization": f"Bearer {api_token}",
    "X-MerchantId": merchant_id,
    "Content-Type": "application/json",
}

response = requests.post(transfer_method_url, headers=headers, json=body)
print(response.json())

# THE CALL ABOVE RETURNS THE FOLLOWING ERROR:
# {
#     "statusCode": -72,
#     "isSuccess": False,
#     "logIdentifier": "76a947c990e24c76b2a62f26ebf48639",
#     "message": "400: Bad request. The inputs supplied to the API are invalid",
#     "errors": {
#         "": [
#             "Invalid state! (Ex. BeneficiaryState allows 2 characters for BeneficiaryCountry US only.)"
#         ]
#     },
# }
# Im currently using "beneficiaryState": "CA" and it still gives the error above. I added beneficiaryCountry and it didn't made a difference

transfer_url = f"{base_url}/transfer/create"
body = {
    "merchantTransactionId": "TX123456",
    "beneficiaryToken": customer_token,
    "autoApprove": True,
    "comments": "Payment for services",
    "dateExpire": "2024-12-31T23:59:59.999Z",
    "destinationAmount": 100.00,
    "destinationCurrency": "USD",
    "destinationType": "Wire",
    "wireProfile": {
        "accountNickName": "John's Wire Account",
        "beneficiaryCountry": "US",
        "accountType1": "personal",
        "beneficiaryAddress1": "123 Elm St",
        "beneficiaryCity": "Los Angeles",
        "beneficiaryState": "CA",
        "beneficiaryZipCode": "90001",
        "bankCountry": "US",
        "routingNumber": "123456789",
        "accountNumber": "987654321",
        "bankName": "Bank of America",
        "bankAddress1": "123 Bank St",
        "bankCity": "New York",
        "bankState": "New York",
        "bankZipCode": "10001",
    },
}
headers = {
    "Authorization": f"Bearer {api_token}",
    "X-MerchantId": merchant_id,
    "Content-Type": "application/json",
}

response = requests.post(transfer_method_url, headers=headers, json=body)

print(response.json())

File: i-payout/recipes/java/swiftWire.java
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.net.URI;
import java.util.Base64;
import org.json.JSONObject;

public class swiftWire {
  private static final String BASE_URL = "https://merchantapi.testewallet.com/api/v1";
  private static final String USERNAME = "<USERNAME>"; // Replace with your username
  private static final String PASSWORD = "<PASSWORD>"; // Replace with your password
  private static final String MERCHANT_ID = "<MERCHANT_ID>"; // Replace with your merchant ID

  public static void main(String[] args) {
    try {
      String apiToken = getApiToken();
      String beneficiaryToken = createBeneficiary(apiToken);
      addWireTransferMethod(apiToken, beneficiaryToken);
      createWireTransfer(apiToken, beneficiaryToken);
    } catch (Exception e) {
      e.printStackTrace();
    }
  }

  private static String getApiToken() throws Exception {
    String authStr = USERNAME + ":" + PASSWORD;
    String encodedAuthStr = Base64.getEncoder().encodeToString(authStr.getBytes());

    HttpClient client = HttpClient.newHttpClient();
    HttpRequest request = HttpRequest.newBuilder()
        .uri(URI.create(BASE_URL + "/authentication/login"))
        .header("accept", "application/json")
        .header("authorization", "Basic " + encodedAuthStr)
        .header("X-MerchantId", MERCHANT_ID)
        .build();

    HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
    JSONObject jsonResponse = new JSONObject(response.body());
    return jsonResponse.getJSONObject("data").getString("token");
  }

  private static String createBeneficiary(String apiToken) throws Exception {
    JSONObject body = new JSONObject();
    body.put("username", "<USERNAME>");
    body.put("firstName", "<FIRST_NAME>");
    body.put("lastName", "<LAST_NAME>");
    body.put("emailAddress", "<EMAIL_ADDRESS>");

    HttpClient client = HttpClient.newHttpClient();
    HttpRequest request = HttpRequest.newBuilder()
        .uri(URI.create(BASE_URL + "/beneficiaries"))
        .header("Authorization", "Bearer " + apiToken)
        .header("X-MerchantId", MERCHANT_ID)
        .header("Content-Type", "application/json")
        .POST(HttpRequest.BodyPublishers.ofString(body.toString()))
        .build();

    HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
    JSONObject jsonResponse = new JSONObject(response.body());
    return jsonResponse.getJSONObject("data").getString("beneficiaryToken");
  }

  private static void addWireTransferMethod(String apiToken, String beneficiaryToken) throws Exception {
    JSONObject body = new JSONObject();
    body.put("accountNickName", "<ACCOUNT_NICKNAME>");
    body.put("beneficiaryCountry", "<BENEFICIARY_COUNTRY>");
    body.put("accountType1", "<ACCOUNT_TYPE_1>");
    body.put("beneficiaryAddress1", "<BENEFICIARY_ADDRESS>");
    body.put("beneficiaryCity", "<BENEFICIARY_CITY>");
    body.put("beneficiaryState", "<BENEFICIARY_STATE>");
    body.put("beneficiaryZipCode", "<BENEFICIARY_ZIP_CODE>");
    body.put("bankCountry", "<BANK_COUNTRY>");
    body.put("accountNumber", "<ACCOUNT_NUMBER>");
    body.put("bankName", "<BANK_NAME>");
    body.put("bankAddress1", "<BANK_ADDRESS>");
    body.put("bankCity", "<BANK_CITY>");
    body.put("bankState", "<BANK_STATE>");
    body.put("bankZipCode", "<BANK_ZIP_CODE>");

    HttpClient client = HttpClient.newHttpClient();
    HttpRequest request = HttpRequest.newBuilder()
        .uri(URI.create(BASE_URL + "/transfermethods/beneficiaries/" + beneficiaryToken + "/wires"))
        .header("Authorization", "Bearer " + apiToken)
        .header("X-MerchantId", MERCHANT_ID)
        .header("Content-Type", "application/json")
        .POST(HttpRequest.BodyPublishers.ofString(body.toString()))
        .build();

    client.send(request, HttpResponse.BodyHandlers.ofString());
  }

  private static void createWireTransfer(String apiToken, String beneficiaryToken) throws Exception {
    JSONObject body = new JSONObject();
    body.put("merchantTransactionId", "<TRANSACTION_ID>");
    body.put("beneficiaryToken", beneficiaryToken);
    body.put("autoApprove", true);
    body.put("comments", "<COMMENTS>");
    body.put("dateExpire", "<EXPIRATION_DATE>");
    body.put("destinationAmount", "<AMOUNT>");
    body.put("destinationCurrency", "<DESTINATION_CURRENCY>");
    body.put("destinationType", "<DESTINATION_TYPE>");

    JSONObject wireProfile = new JSONObject();
    wireProfile.put("accountNickName", "<ACCOUNT_NICKNAME>");
    wireProfile.put("beneficiaryCountry", "<BENEFICIARY_COUNTRY>");
    wireProfile.put("accountType1", "<ACCOUNT_TYPE_1>");
    wireProfile.put("beneficiaryAddress1", "<BENEFICIARY_ADDRESS>");
    wireProfile.put("beneficiaryCity", "<BENEFICIARY_CITY>");
    wireProfile.put("beneficiaryState", "<BENEFICIARY_STATE>");
    wireProfile.put("beneficiaryZipCode", "<BENEFICIARY_ZIP_CODE>");
    wireProfile.put("bankCountry", "<BANK_COUNTRY>");
    wireProfile.put("routingNumber", "<ROUTING_NUMBER>");
    wireProfile.put("accountNumber", "<ACCOUNT_NUMBER>");
    wireProfile.put("bankName", "<BANK_NAME>");
    wireProfile.put("bankAddress1", "<BANK_ADDRESS>");
    wireProfile.put("bankCity", "<BANK_CITY>");
    wireProfile.put("bankState", "<BANK_STATE>");
    wireProfile.put("bankZipCode", "<BANK_ZIP_CODE>");
    body.put("wireProfile", wireProfile);

    HttpClient client = HttpClient.newHttpClient();
    HttpRequest request = HttpRequest.newBuilder()
        .uri(URI.create(BASE_URL + "/transfers"))
        .header("Authorization", "Bearer " + apiToken)
        .header("X-MerchantId", MERCHANT_ID)
        .header("Content-Type", "application/json")
        .POST(HttpRequest.BodyPublishers.ofString(body.toString()))
        .build();

    HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
    System.out.println(response.body());
  }
}
